### ISSUE TITLE
minetest.serialize truncates big numbers on some systems (Raspberry Pi for instance)

### ISSUE THREAD

**Issue by pyrollo** [minetest/minetest#7925]
##### Issue type
 - Bug report
 
##### Minetest version
Minetest Server Version 0.4.17-1 compiled from source

##### OS / Hardware
Operating system: Raspbian Lite (Debian 9.6)
Machine: Raspberry Pi 2 / Raspberry Pi 3

##### Summary
Using `minetest.serialize` and `minetest.deserialize` on Raspberry Pi with long numbers (> 32 bits) truncates the number to a max 32 bit number.

##### Steps to reproduce
A good case is serializing a table storing a position hash. This is a simplified extract of code storing a position hash in an entity static data:
```
local pos = {...}
local staticdata = minetest.serialize({pos=minetest.hash_node_position(pos)})
...
local data = minetest.deserialize(staticdata)
```
Retrieved pos hash is not identical before and after.

Some more information can be found in the original issue involving position loss: https://github.com/pyrollo/display_modpack/issues/19

What I suspect is that the "%q" `string.format` flag is not working the same way on this system than on x64 systems.

Unfortunately I do not have any Rapberry and can not investigate further.

**Comment by paramat** [comment]
Seems low priority, why use numbers larger than 32 bits anyway? That's pretty big.

**Comment by pyrollo** [comment]
Yes it's big but that's what `minetest.hash_node_position` gives (48 bits integers).
Actually, it is not that important to fix it, but it's important to be aware that it's not possible to serialize hash_node_position result on some systems .
We could say that issue is opened for the records.

**Comment by SmallJoker** [comment]
Would it be possible to try this change?
https://github.com/minetest/minetest/blob/ff12630bc966ef6d72f4c6d07c59d952fd44dfb7/builtin/common/serialize.lua#L128
to:
```Lua
return string.format("%.30g", x)
```

Source: https://gitlab.com/VanessaE/pipeworks/commit/57fc8c67

**Comment by pyrollo** [comment]
It seems to be a good solution. Now, only missing thing is ... a Raspberry owner to test it :)

**Comment by sbrl** [comment]
What would testing it involve?

**Comment by JDCodeIt** [comment]
No going over to the Pi 3B, yet, but here is the result in 64bit linux with lua 5.1. Appears there is a precision problem at digit 17. I see notes about later lua versions supporting true integers and at higher precision.
```lua
d = 12345678901234567
print(d)
print(string.format("%.30g",d))
e = 12345678901234567890
print(e)
print(string.format("%.30g",e))
```
Results:
1.2345678901235e+16
12345678901234568
1.2345678901235e+19
12345678901234567168

**Comment by HybridDog** [comment]
`minetest.serialize` currently truncates the double variable I think:
```lua
local x = math.random()
print(x == minetest.deserialize(minetest.serialize(x)))
```
This gives false for me (I hope I didn't write a typo).
The double is converted to a decimal number string.

For a reversible serialization, maybe `%a` could help:
```lua
local x = math.random()
print(x == tonumber(("%a"):format(x)))
```

**Comment by sbrl** [comment]
@HybridDog: Output on a Raspberry Pi 3B+ (unfortunately I'm not yet certain how to run Lua code directly in the context of a Minetest server):

```
$ lua
Lua 5.1.5  Copyright (C) 1994-2012 Lua.org, PUC-Rio
> local x = math.random()
> print(x == minetest.deserialize(minetest.serialize(x)))
stdin:1: attempt to index global 'minetest' (a nil value)
stack traceback:
	stdin:1: in main chunk
	[C]: ?
> local x = math.random()
> print(x == tonumber(("%a"):format(x)))
stdin:1: invalid option '%a' to 'format'
stack traceback:
	[C]: in function 'format'
	stdin:1: in main chunk
	[C]: ?
```

Sorry for the late reply :P

**Comment by HybridDog** [comment]
> unfortunately I'm not yet certain how to run Lua code directly in the context of a Minetest server

You can use the [luacmd](https://github.com/prestidigitator/minetest-mod-luacmd) mod, or simply add the code to the code of some mod which is enabled on the server.

> `invalid option '%a' to 'format'`

You need to use Lua 5.2 or LuaJIT because [`%a` is not yet available in Lua 5.1](https://stackoverflow.com/a/51222710).

**Comment by sbrl** [comment]
Got it, @HybridDog! Looks like my Minetest 5.2.0 server has the `/lua` command already (again, on the Raspberry Pi 3B+). Transcript:

```
] /lua local x = math.random(); print(x == minetest.deserialize(minetest.serialize(x))); local y = math.random(); print(x == tonumber(("%a"):format(x)))
false
true
```

**Comment by HybridDog** [comment]
Unfortunately, `%a` is not available in Lua 5.1.5, which can be used in Minetest.
`%.30g` could work.
